submitit/local/local.py

```diff
@@ -21,27 +21,33 @@
 
 LOCAL_REQUEUE_RETURN_CODE = 144
 
+PROCESSES = {}
+
 
 class LocalJob(core.Job[R]):
     def __init__(
         self,
-        folder: Union[Path, str],
+        folder: tp.Union[Path, str],
         job_id: str,
-        tasks: Sequence[int] = (0,),
-        process: Optional["subprocess.Popen['bytes']"] = None,
+        tasks: tp.Sequence[int] = (0,),
+        process: tp.Optional["subprocess.Popen['bytes']"] = None,
     ) -> None:
         super().__init__(folder, job_id, tasks)
         self._cancel_at_deletion = False
-        self._process = process
         # downcast sub-jobs to get proper typing
-        self._sub_jobs: Sequence["LocalJob[R]"] = self._sub_jobs
-        for sjob in self._sub_jobs:
-            sjob._process = process
+        self._sub_jobs: tp.Sequence["LocalJob[R]"] = self._sub_jobs
+        # set process (to self and subjobs)
+        if process is not None:
+            for sjob in [self] + list(self._sub_jobs):
+                PROCESSES[sjob.job_id] = process
```

Comment:
This prevents the processes to ever get garbage collected.
Could it be a problem if you have a lot of those job ? Especially because those process object may have a full copy of the stdout/stderr.

The thing is that if you want to pickle/unpickle you need the process to survive this object.
Maybe you could implement `__getstate__` and `__setstate__` such as when `__getstate__` is called, you copy `self._process` to `PROCESSES` and on `__setstate__` you read back `self._process` from `PROCESSES`.

Reply:
then we'd need to remove the process from the dict when reloading right? But since subtasks share the same process that would not work. Or if we load the pickle twice it also fails.
Then again we could also not remove it indeed.
I was thinking of weakrefs as well but that wont work

On a side note when the job is garbage collected, the process is removed from the dict if it is finished, so I don't expect it would be that much of a mess... then again you never know indeed

Reply:
I ended up implemented adding processes in the dict only at getstate as you proposed, and deleting at job deletion if finished, that should cover as much use cases as possible with very limited overhead